src/parser.rs

```diff
@@ -0,0 +1,589 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2021, tree-sitter authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+use std::fmt::Display;
+use std::iter::Peekable;
+use std::str::Chars;
+
+use regex::Regex;
+use thiserror::Error;
+use tree_sitter::Language;
+use tree_sitter::Query;
+use tree_sitter::QueryError;
+
+use crate::ast;
+use crate::Context;
+use crate::Identifier;
+
+impl ast::File {
+    /// Parses a graph DSL file, adding its content to an existing `File` instance.
+    pub fn parse(&mut self, ctx: &mut Context, content: &str) -> Result<(), ParseError> {
+        Parser::new(ctx, content).parse_into_file(self)
+    }
+}
+
+/// An error that can occur while parsing a graph DSL file
+#[derive(Debug, Error)]
+pub enum ParseError {
+    #[error("Expected '{0}' at {1}")]
+    ExpectedToken(&'static str, Location),
+    #[error("Expected variable name at {0}")]
+    ExpectedVariable(Location),
+    #[error("Invalid regular expression /{0}/ at {1}")]
+    InvalidRegex(String, Location),
+    #[error("Expected integer constant in regex capture at {0}")]
+    InvalidRegexCapture(Location),
+    // TODO: The positions in the wrapped QueryError will be incorrect, since they will count the
+    // row/column from the start of the query, not from the start of the file.
+    #[error("Invalid query pattern: {}", _0.message)]
+    QueryError(#[from] QueryError),
+    #[error("Undefined query capture '{0}' at {1}")]
+    UndefinedCapture(String, Location),
+    #[error("Unexpected character '{0}' in {1} at {2}")]
+    UnexpectedCharacter(char, &'static str, Location),
+    #[error("Unexpected end of file at {0}")]
+    UnexpectedEOF(Location),
+    #[error("Unexpected keyword '{0}' at {1}")]
+    UnexpectedKeyword(String, Location),
+    #[error("Unexpected literal '#{0}' at {1}")]
+    UnexpectedLiteral(String, Location),
+}
+
+/// The location of a graph DSL entity within its file
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Location {
+    pub row: usize,
+    pub column: usize,
+}
+
+impl Location {
+    fn advance(&mut self, ch: char) {
+        if ch == '\n' {
+            self.row += 1;
+            self.column = 0;
+        } else {
+            self.column += 1;
+        }
+    }
+}
+
+impl Display for Location {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "({}, {})", self.row + 1, self.column + 1)
+    }
+}
+
+struct Parser<'a> {
```

Comment:
Is the rest of this generated code, or should someone (probably not me 😛) review it exhaustively?

Reply:
It's not generated, but it is copy-pasted over from the internal repo where it was reviewed previously...  :smile: